File: web/src/crypto/group.ts
// Group / room crypto — the browser counterpart of tui/internal/crypto/group.go.
// Every operation here is byte-compatible with the Go implementation so a web
// client and a TUI client can share a room transparently.
//
//   - Room key wrap/unwrap : nacl/box  (X25519 + XSalsa20-Poly1305)
//   - Message AEAD         : XChaCha20-Poly1305 (24-byte random nonce, IETF)
//   - Epoch ratchet        : HKDF-SHA256
//
// The AEAD additional data is the 8-byte big-endian epoch, and the message
// signature covers protocol/signing.ts SigningBytes — both identical to Go.

import nacl from "tweetnacl";
import { xchacha20poly1305 } from "@noble/ciphers/chacha";
import { hkdf } from "@noble/hashes/hkdf";
import { sha256 } from "@noble/hashes/sha256";
import { randomBytes } from "./random";
import { signingBytes, u64be, utf8 } from "./signing";
import type { Identity } from "./identity";

export interface RoomKey {
  epoch: number;
  key: Uint8Array; // 32 bytes
}

// HKDF info string for the forward-secret epoch ratchet. Must match Go's
// "netherchat/epoch-ratchet/v1" exactly.
const RATCHET_INFO = utf8("netherchat/epoch-ratchet/v1");

/** Fresh random room key. Used by the first member (epoch 0) and on member add. */
export function newRoomKey(epoch = 0): RoomKey {
  return { epoch, key: randomBytes(32) };
}

/**
 * Advance the room key by one epoch via HKDF (salt nil -> zero-filled, matching
 * Go's hkdf.New with a nil salt). Deterministic, so every member that applies it
 * lands on the same next key with no key exchange — used for /vanish.
 */
export function ratchet(rk: RoomKey): RoomKey {
  const next = hkdf(sha256, rk.key, undefined, RATCHET_INFO, 32);
  return { epoch: rk.epoch + 1, key: next };
}

/** Seal a room key for a recipient's X25519 public key (nacl/box). */
export function wrapRoomKey(
  id: Identity,
  rk: RoomKey,
  recipientKX: Uint8Array,
): { nonce: Uint8Array; wrapped: Uint8Array } {
  const nonce = randomBytes(24);
  const wrapped = nacl.box(rk.key, nonce, recipientKX, id.kxSecret);
  return { nonce, wrapped };
}

/** Open a wrapped room key, verifying it came from senderKX. */
export function unwrapRoomKey(
  id: Identity,
  epoch: number,
  nonce: Uint8Array,
  wrapped: Uint8Array,
  senderKX: Uint8Array,
): RoomKey {
  if (nonce.length !== 24) throw new Error(`wrap nonce wrong length: ${nonce.length}`);
  const opened = nacl.box.open(wrapped, nonce, senderKX, id.kxSecret);
  if (!opened) throw new Error("room key unwrap failed: box authentication error");
  if (opened.length !== 32) throw new Error(`unwrapped room key wrong length: ${opened.length}`);
  return { epoch, key: opened };
}

export interface SealedMessage {
  nonce: Uint8Array;
  ciphertext: Uint8Array;
  signature: Uint8Array;
}

/**
 * Encrypt plaintext under the room key (XChaCha20-Poly1305, epoch as AAD) and
 * sign the result with the sender's Ed25519 key.
 */
export function sealMessage(
  id: Identity,
  rk: RoomKey,
  roomID: string,
  fromID: string,
  plaintext: Uint8Array,
): SealedMessage {
  const nonce = randomBytes(24);
  const aead = xchacha20poly1305(rk.key, nonce, u64be(rk.epoch));
  const ciphertext = aead.encrypt(plaintext);
  const signature = nacl.sign.detached(
    signingBytes(roomID, fromID, rk.epoch, nonce, ciphertext),
    id.signSecret,
  );
  return { nonce, ciphertext, signature };
}

/** The result of openMessage: the plaintext, plus whether it was authenticated. */
export interface OpenedMessage {
  plaintext: Uint8Array;
  /**
   * True only if the frame carried a signature AND it verified. False means the
   * frame arrived unsigned — a legacy/pre-v3 sender, or a relay that stripped
   * `sig` — so the sender is attributed by routing metadata alone, not proven.
   * A signature that is present but INVALID never gets here: it throws.
   */
  signed: boolean;
}

/**
 * Decrypt a message and report whether it carried a valid signature — the
 * browser counterpart of Go's OpenMessage, which returns (plaintext, signed,
 * err). If a signature is present it is verified BEFORE decrypting (so a forged
 * ciphertext never touches the AEAD); an invalid signature throws. An empty
 * signature is accepted as "unsigned" (legacy / pre-v3, signed=false) and
 * decrypted without verification. Throws on decrypt failure or epoch mismatch.
 */
export function openMessage(
  rk: RoomKey,
  senderSignPub: Uint8Array,
  roomID: string,
  fromID: string,
  epoch: number,
  nonce: Uint8Array,
  ciphertext: Uint8Array,
  signature: Uint8Array,
): OpenedMessage {
  let signed = false;
  if (signature.length > 0) {
    if (senderSignPub.length !== 32) throw new Error("sender signing key invalid");
    if (
      !nacl.sign.detached.verify(
        signingBytes(roomID, fromID, epoch, nonce, ciphertext),
        signature,
        senderSignPub,
      )
    ) {
      throw new Error("message signature verification failed");
    }
    signed = true;
  }
  if (rk.epoch !== epoch) {
    throw new Error(`epoch mismatch: holding key for epoch ${rk.epoch}, message is epoch ${epoch}`);
  }
  const aead = xchacha20poly1305(rk.key, nonce, u64be(epoch));
  return { plaintext: aead.decrypt(ciphertext), signed };
}

File: tui/internal/crypto/group.go
package crypto

import (
	"crypto/ed25519"
	"crypto/rand"
	"crypto/sha256"
	"encoding/binary"
	"errors"
	"fmt"
	"io"

	"github.com/salehkreiner/netherchat/protocol"
	"golang.org/x/crypto/chacha20poly1305"
	"golang.org/x/crypto/hkdf"
	"golang.org/x/crypto/nacl/box"
)

// RoomKey is the symmetric key shared by every member of a room for one epoch.
// Messages within an epoch are sealed under Key with XChaCha20-Poly1305. A new
// epoch begins on membership change, /vanish, or rotation cadence; see the
// epoch model in ARCHITECTURE_DECISION.md §3.
type RoomKey struct {
	Epoch uint64
	Key   [32]byte
}

// NewRoomKey mints a fresh random room key at the given epoch. Used by the first
// member of a room (epoch 0) and whenever a member is ADDED — a random rekey
// ensures a newcomer cannot derive keys for epochs before it joined.
func NewRoomKey(epoch uint64) (RoomKey, error) {
	rk := RoomKey{Epoch: epoch}
	if _, err := rand.Read(rk.Key[:]); err != nil {
		return RoomKey{}, fmt.Errorf("generate room key: %w", err)
	}
	return rk, nil
}

// Ratchet derives the next epoch's key from the current one via HKDF and returns
// it. This provides forward secrecy at epoch granularity: once the caller
// discards (Zero) the old key, messages from the previous epoch cannot be
// decrypted even if the new key is later compromised. Used for rotation that
// preserves the existing membership (e.g. /vanish, time cadence), as opposed to
// NewRoomKey which is used when adding members.
func (rk RoomKey) Ratchet() (RoomKey, error) {
	r := hkdf.New(sha256.New, rk.Key[:], nil, []byte("netherchat/epoch-ratchet/v1"))
	next := RoomKey{Epoch: rk.Epoch + 1}
	if _, err := io.ReadFull(r, next.Key[:]); err != nil {
		return RoomKey{}, fmt.Errorf("ratchet room key: %w", err)
	}
	return next, nil
}

// Zero wipes the key material from memory. Best-effort (Go's GC may have copied
// it), but it shortens the window in which a deleted key sits in process memory.
func (rk *RoomKey) Zero() {
	for i := range rk.Key {
		rk.Key[i] = 0
	}
}

// WrapRoomKey seals rk for a recipient's X25519 public key using nacl/box
// (authenticated X25519 + XSalsa20-Poly1305). The box is authenticated by the
// sender's X25519 private key, so the recipient learns which member distributed
// the key. The returned nonce/wrapped blob are opaque to the relay server.
func (id *Identity) WrapRoomKey(rk RoomKey, recipientKX [32]byte) (nonce, wrapped []byte, err error) {
	var n [24]byte
	if _, err := rand.Read(n[:]); err != nil {
		return nil, nil, fmt.Errorf("nonce: %w", err)
	}
	sealed := box.Seal(nil, rk.Key[:], &n, &recipientKX, &id.KXPriv)
	return n[:], sealed, nil
}

// UnwrapRoomKey opens a key blob produced by WrapRoomKey, verifying it came from
// senderKX. Returns the room key at the given epoch.
func (id *Identity) UnwrapRoomKey(epoch uint64, nonce, wrapped []byte, senderKX [32]byte) (RoomKey, error) {
	if len(nonce) != 24 {
		return RoomKey{}, fmt.Errorf("wrap nonce wrong length: got %d, want 24", len(nonce))
	}
	var n [24]byte
	copy(n[:], nonce)
	opened, ok := box.Open(nil, wrapped, &n, &senderKX, &id.KXPriv)
	if !ok {
		return RoomKey{}, errors.New("room key unwrap failed: box authentication error")
	}
	if len(opened) != 32 {
		return RoomKey{}, fmt.Errorf("unwrapped room key wrong length: got %d, want 32", len(opened))
	}
	rk := RoomKey{Epoch: epoch}
	copy(rk.Key[:], opened)
	return rk, nil
}

// ErrBadSignature is returned by OpenMessage when a present signature fails
// verification. Such a message MUST be rejected (and its body never shown).
var ErrBadSignature = errors.New("message signature verification failed")

// SealMessage encrypts plaintext under the room key with XChaCha20-Poly1305 and
// signs the result with the sender's Ed25519 identity key. The signature binds
// the room id, sender id, and epoch (§3.3); the epoch is also AEAD additional
// data. Returns the wire fields for a protocol.Message.
func (id *Identity) SealMessage(rk RoomKey, roomID, fromID string, plaintext []byte) (nonce, ciphertext, signature []byte, err error) {
	aead, err := chacha20poly1305.NewX(rk.Key[:])
	if err != nil {
		return nil, nil, nil, fmt.Errorf("init aead: %w", err)
	}
	nonce = make([]byte, chacha20poly1305.NonceSizeX) // 24 bytes; random is safe with XChaCha
	if _, err := rand.Read(nonce); err != nil {
		return nil, nil, nil, fmt.Errorf("nonce: %w", err)
	}
	ciphertext = aead.Seal(nil, nonce, plaintext, epochAD(rk.Epoch))
	signature, err = id.Sign(protocol.SigningBytes(roomID, fromID, rk.Epoch, nonce, ciphertext))
	if err != nil {
		return nil, nil, nil, fmt.Errorf("sign message: %w", err)
	}
	return nonce, ciphertext, signature, nil
}

// OpenMessage decrypts a message and reports whether it carried a valid
// signature. If sig is non-empty it is verified BEFORE decryption (a forged
// ciphertext is rejected without touching the AEAD) against senderSignPub over
// the room-bound signing bytes; an invalid signature returns ErrBadSignature. If
// sig is empty the message is accepted as UNSIGNED (signed=false) and decrypted
// without verification — pre-v3 / legacy frames interoperate this way.
func OpenMessage(rk RoomKey, senderSignPub ed25519.PublicKey, roomID, fromID string, epoch uint64, nonce, ciphertext, sig []byte) (plaintext []byte, signed bool, err error) {
	if len(sig) > 0 {
		if len(senderSignPub) != ed25519.PublicKeySize {
			return nil, false, errors.New("sender signing key invalid")
		}
		if !ed25519.Verify(senderSignPub, protocol.SigningBytes(roomID, fromID, epoch, nonce, ciphertext), sig) {
			return nil, false, ErrBadSignature
		}
		signed = true
	}
	if rk.Epoch != epoch {
		return nil, signed, fmt.Errorf("epoch mismatch: holding key for epoch %d, message is epoch %d", rk.Epoch, epoch)
	}
	aead, err := chacha20poly1305.NewX(rk.Key[:])
	if err != nil {
		return nil, signed, fmt.Errorf("init aead: %w", err)
	}
	plaintext, err = aead.Open(nil, nonce, ciphertext, epochAD(epoch))
	if err != nil {
		return nil, signed, fmt.Errorf("decrypt: %w", err)
	}
	return plaintext, signed, nil
}

// epochAD is the AEAD additional-authenticated-data for a message: the epoch,
// big-endian. Binding the epoch prevents a ciphertext from being reinterpreted
// under a different epoch.
func epochAD(epoch uint64) []byte {
	var b [8]byte
	binary.BigEndian.PutUint64(b[:], epoch)
	return b[:]
}
